docs: grammar fixes and minor clarifications (#9150)
facebook/docusaurus: website/docs/deployment.mdx

:::

> You can deploy your site to static site hosting services such as [Vercel](https://vercel.com/), [GitHub Pages](https://pages.github.com/), [Netlify](https://www.netlify.com/), [Render](https://render.com/docs/static-sites) and [Surge](https://surge.sh/help/getting-started-with-surge).

Reviewer: ```suggestion
You can deploy your site to static site hosting services such as [Vercel](https://vercel.com/), [GitHub Pages](https://pages.github.com/), [Netlify](https://www.netlify.com/), [Render](https://render.com/docs/static-sites), and [Surge](https://surge.sh/help/getting-started-with-surge).
```

## Using environment variables {#using-environment-variables}

> Putting potentially sensitive information in the environment is common practice. However, in a typical Docusaurus website, the `docusaurus.config.js` file is the only interface to the Node.js environment (see [our architecture overview](advanced/architecture.mdx)), while everything else (MDX pages, React components, etc) are client side and do not have direct access to the `process` global variable. In this case, you can consider using [`customFields`](api/docusaurus.config.js.mdx#customFields) to pass environment variables to the client side.

Reviewer: ```suggestion
Putting potentially sensitive information in the environment is common practice. However, in a typical Docusaurus website, the `docusaurus.config.js` file is the only interface to the Node.js environment (see [our architecture overview](advanced/architecture.mdx)), while everything else (MDX pages, React components, etc。) are client side and do not have direct access to the `process` global variable. In this case, you can consider using [`customFields`](api/docusaurus.config.js.mdx#customFields) to pass environment variables to the client side.
```

There are a few common hosting options:

- [Self hosting](#self-hosting) with an HTTP server like Apache2 or Nginx.
- Jamstack providers (e.g. [Netlify](#deploying-to-netlify) and [Vercel](#deploying-to-vercel)). We will use them as references, but the same reasoning can apply to other providers.
- [GitHub Pages](#deploying-to-github-pages) (by definition, it is also Jamstack, but we compare it separately).

If you are unsure of which one to choose, ask the following questions:

<details>

<summary>

>   How much time and money am I willing to invest in this?

Reviewer 2: Fine 👍 

```suggestion
  How many resources (e.g. money, person-hours, etc.) am I willing to invest in this?
```

There are a few common hosting options:

> - [Self hosting](#self-hosting) with an HTTP server like Apache2 or Nginx

Reviewer: ```suggestion
- [Self hosting](#self-hosting) with an HTTP server like Apache2 or Nginx.
```

- Jamstack providers (e.g. [Netlify](#deploying-to-netlify) and [Vercel](#deploying-to-vercel)). We will use them as references, but the same reasoning can apply to other providers.

> - [GitHub Pages](#deploying-to-github-pages) (By definition, it is also Jamstack, but we compare it separately).

Reviewer: ```suggestion
- [GitHub Pages](#deploying-to-github-pages) (by definition, it is also Jamstack, but we compare it separately).
```

If you are unsure of which one to choose, ask the following questions:

<details>

<summary>

>   How much time and money am I willing to invest in this?

Reviewer: "Time and money" is non-exhaustive, so I prefer the wording before, possibly adding "etc." or "e.g.".

Author: How about:

How many resources (e.g. money, person-hours, etc.) am I willing to invest in this?

Reviewer: Sure. I usually avoid both "e.g." and "etc." so you can choose to omit either one.

Author: Good point; I'll take out the e.g.

</summary>

- 🔴 Self-hosting requires experience in networking as well as Linux and web server administration. It's the most difficult option, and would require the most time to manage successfully. Expense-wise, cloud services are almost never free, and purchasing/deploying an onsite server can be even more costly.
- 🟢 Jamstack providers can help you set up a working website in almost no time and offers features like server-side redirects that are easily configurable. Many providers offer generous build-time quotas even for free plans that you would almost never exceed. However, free plans have limits, and you would need to pay once you hit those limits. Check the pricing page of your provider for details.

> - 🟡 The GitHub Pages deployment workflow can be tedious to set up (Evidence: see the length of [Deploying to GitHub Pages](#deploying-to-github-pages)!). However, this service (including build and deployment) is always free for public repositories, and we have detailed instructions to help you make it work.

Reviewer: The parentheses here enclose a complete sentence (technically not, but semantically so), so it's right to put the period before.

```suggestion
- 🟡 The GitHub Pages deployment workflow can be tedious to set up. (Evidence: see the length of [Deploying to GitHub Pages](#deploying-to-github-pages)!) However, this service (including build and deployment) is always free for public repositories, and we have detailed instructions to help you make it work.
```

</details>

<details>

<summary>How much server-side customization do I need?</summary>

- 🟢 With self-hosting, you have access to the entire server's configuration. You can configure the virtual host to serve different content based on the request URL, you can do complicated server-side redirects, you can implement authentication, and so on. If you need a lot of server-side features, self-host your website.
- 🟡 Jamstack usually offers some server-side configuration (e.g. URL formatting (trailing slashes), server-side redirects, etc.).
- 🔴 GitHub Pages doesn't expose server-side configuration besides enforcing HTTPS and setting CNAME records.

</details>

<details>

> <summary>Do I need collaboration-friendly deployment workflows?</summary>

Reviewer: Nice 👍

> - 🟡 Self-hosted services can leverage continous deployment functionality like Netlify, but more heavy-lifting is involved. Usually, you would designate a specific person to manage the deployment, and the workflow wouldn't be very git-based as opposed to the other two options.

Reviewer: ```suggestion
- 🟡 Self-hosted services can leverage continuous deployment functionality like Netlify, but more heavy-lifting is involved. Usually, you would designate a specific person to manage the deployment, and the workflow wouldn't be very git-based as opposed to the other two options.
```

In the original wording, we are trying to suggest that you can "self-implement Netlify" because you are really not using any CI/CD. I don't have strong opinions if you want to suggest incorporating Netlify, but just for you to consider

Author: Maybe this would be more accurate:

You can achieve similar functionality as Netlify with self-hosting services, but more heavy-lifting is involved . . . . [text continues]

I can't believe I misspelled "continuous." That's a little embarrassing.  :-O

- 🟢 Netlify and Vercel have deploy previews for every pull request, which is useful for a team to review work before merging to production. You can also manage a team with different member access to the deployment.

> - 🔴 GitHub Pages cannot do deploy previews in a non-convoluted way. One repo can only be associated with one site deployment. On the other hand, you can control who has write access to the site's deployment.

Reviewer: It was meant to be yellow, because it's not "absolutely terrible" ^^

Author: Gotcha. :-)

:::warning

> In the following sections, we will introduce a few common hosting providers and how they should be configured to deploy Docusaurus sites most efficiently. Docusaurus is not affiliated with any of these services, and this information is provided for convenience only. However, if you see outdated content, PRs are welcome.

Reviewer: > Some of the writeups are provided by external contributors.
> The documentation may not be up-to-date: recent changes in their API may not be reflected on our side.

These are important caveats; why delete them?

Author: I was trying to be concise by covering them under the "this info is provided for convenience only/use at your own risk" umbrella, but if we want to be less vague, maybe this (changes in bold italics):

In the following sections, we will introduce a few common hosting providers and how they should be configured to deploy Docusaurus sites most efficiently. Docusaurus is not affiliated with any of these services, and this information is provided for convenience only. _**Some of the write-ups are provided by third-parties, and recent API changes may not be reflected on our side. If**_  you see outdated content, PRs are welcome.

Reviewer: That is okay for me ^^

Author: Great, I'll make the change.

To deploy your Docusaurus project with a [Vercel for Git Integration](https://vercel.com/docs/concepts/git), make sure it has been pushed to a Git repository.

Import the project into Vercel using the [Import Flow](https://vercel.com/import/git). During the import, you will find all relevant options preconfigured for you; however, you can choose to change any of these [options](https://vercel.com/docs/build-step#build-&-development-settings).

After your project has been imported, all subsequent pushes to branches will generate [Preview Deployments](https://vercel.com/docs/platform/deployments#preview), and all changes made to the [Production Branch](https://vercel.com/docs/git-integrations#production-branch) (usually "main" or "master") will result in a [Production Deployment](https://vercel.com/docs/platform/deployments#production).

## Deploying to GitHub Pages {#deploying-to-github-pages}

Docusaurus provides an easy way to publish to [GitHub Pages](https://pages.github.com/), which comes free with every GitHub repository.

### Overview {#github-pages-overview}

> Usually, there are two repositories (at least two branches) involved in a publishing process: the branch containing the source files, and the branch containing the build output to be served with GitHub Pages. In the following tutorial, they will be referred to as **"source"** and **"deployment,"** respectively.

Reviewer: We don't use the American style here.

```suggestion
Usually, there are two repositories (at least two branches) involved in a publishing process: the branch containing the source files, and the branch containing the build output to be served with GitHub Pages. In the following tutorial, they will be referred to as **"source"** and **"deployment"**, respectively.
```

:::

> GitHub Pages picks up deploy-ready files (the output from `docusaurus build`) from the default branch (`master` / `main`, usually) or the `gh-pages` branch, and either from the root or the `/docs` folder. You can configure that through `Settings > Pages` in your repository. This branch will be called the "deployment branch."

Reviewer: ```suggestion
GitHub Pages picks up deploy-ready files (the output from `docusaurus build`) from the default branch (`master` / `main`, usually) or the `gh-pages` branch, and either from the root or the `/docs` folder. You can configure that through `Settings > Pages` in your repository. This branch will be called the "deployment branch".
```

3. Copy the key to clipboard with `pbcopy < ~/.ssh/id_rsa.pub` and paste it as a [deploy key](https://developer.github.com/v3/guides/managing-deploy-keys/#deploy-keys) in the deployment repository. Copy the file content if the command line doesn't work for you. Check the box for `Allow write access` before saving your deployment key.
4. You'll need your private key as a [GitHub secret](https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets) to allow Docusaurus to run the deployment for you.
5. Copy your private key with `pbcopy < ~/.ssh/id_rsa` and paste a GitHub secret with the name `GH_PAGES_DEPLOY` on your source repository. Copy the file content if the command line doesn't work for you. Save your secret.
6. Create your [documentation workflow file](https://help.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow#creating-a-workflow-file) in `.github/workflows/`. In this example it's `deploy.yml`.

> At this point, you should have:
>   - the source repo with the GitHub workflow set with the private SSH key as the GitHub Secret
>   - your deployment repo set with the public SSH key in GitHub Deploy Keys

Reviewer: ```suggestion
At this point, you should have:

- the source repo with the GitHub workflow set with the private SSH key as the GitHub Secret, and
- your deployment repo set with the public SSH key in GitHub Deploy Keys.
```

<summary>Site not deployed properly?</summary>

> After pushing to main, if you don't see your site published at the desired location (for example, it says "There isn't a GitHub Pages site here," or it's showing your repo's README.md file), try the following:

Reviewer: ```suggestion
After pushing to main, if you don't see your site published at the desired location (for example, it says "There isn't a GitHub Pages site here", or it's showing your repo's README.md file), try the following:
```

- Wait about three minutes and refresh. It may take a few minutes for GitHub pages to pick up the new files.
- Check your repo's landing page for a little green tick next to the last commit's title, indicating the CI has passed. If you see a cross, it means the build or deployment failed, and you should check the log for more debugging information.

> - Click on the tick and make sure you see a "Deploy to GitHub Pages" workflow. Names like "pages build and deployment / deploy" are GitHub's default workflows, indicating your custom deployment workflow failed to be triggered at all. Make sure the YAML files are placed under the `.github/workflows` folder, and that the trigger condition is set correctly (e.g., if your default branch is "master" instead of "main," you need to change the `on.push` property).

Reviewer: ```suggestion
- Click on the tick and make sure you see a "Deploy to GitHub Pages" workflow. Names like "pages build and deployment / deploy" are GitHub's default workflows, indicating your custom deployment workflow failed to be triggered at all. Make sure the YAML files are placed under the `.github/workflows` folder, and that the trigger condition is set correctly (e.g., if your default branch is "master" instead of "main", you need to change the `on.push` property).
```

> - Make sure the "Source" (which is the source for the _deployment_ files, not "source" as in our terminology) is set to "gh-pages" + "/ (root)" Under your repo's Settings > Pages (We are using `gh-pages` as the deployment branch).

Reviewer: ```suggestion
- Under your repo's Settings > Pages, make sure the "Source" (which is the source for the _deployment_ files, not "source" as in our terminology) is set to "gh-pages" + "/ (root)", since we are using `gh-pages` as the deployment branch.
```

I prefer the words to follow the steps users take (i.e. they first navigate to the tab and _then_ check the settings)

If you are using a custom domain:

> - Verify that you have the correct DNS records set up if you're using a custom domain. See [GitHub pages documentation on configuring custom domains](https://docs.github.com/en/pages/configuring-a-custom-domain-for-your-github-pages-site/about-custom-domains-and-github-pages). Also, please be aware that it may take up to 24 hours for DNS changes to propegate through the internet.

Reviewer: ```suggestion
- Verify that you have the correct DNS records set up if you're using a custom domain. See [GitHub pages documentation on configuring custom domains](https://docs.github.com/en/pages/configuring-a-custom-domain-for-your-github-pages-site/about-custom-domains-and-github-pages). Also, please be aware that it may take up to 24 hours for DNS changes to propagate through the internet.
```
